File: ios/IndeRun/Sources/IndeRunCore/Registry.swift
import Foundation

public final class ProviderRegistry: @unchecked Sendable {
    private let lock = NSRecursiveLock()
    private var providers: [String: any ProviderAdapter] = [:]
    
    public init() {}
    
    public func register(_ provider: any ProviderAdapter) throws {
        lock.lock()
        defer { lock.unlock() }
        
        let descriptor = provider.describe()
        guard !descriptor.id.isEmpty else {
            throw createInternal(message: "Provider must have a non-empty ID.")
        }
        
        if providers[descriptor.id] != nil {
            throw createInternal(message: "Provider with ID \"\(descriptor.id)\" is already registered.")
        }
        
        providers[descriptor.id] = provider
    }
    
    public func get(id: String) -> (any ProviderAdapter)? {
        lock.lock()
        defer { lock.unlock() }
        return providers[id]
    }
    
    public func list() -> [any ProviderAdapter] {
        lock.lock()
        defer { lock.unlock() }
        return Array(providers.values)
    }
    
    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        providers.removeAll()
    }
}

File: android/inderun-core/src/main/kotlin/com/independo/inderun/core/Registry.kt
package com.independo.inderun.core

import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class ProviderRegistry {
    private val lock = ReentrantLock()
    private val providers = linkedMapOf<String, ProviderAdapter>()

    fun register(provider: ProviderAdapter) {
        lock.withLock {
            val descriptor = provider.describe()
            require(descriptor.id.isNotBlank()) { "Provider must have a non-empty ID." }
            check(!providers.containsKey(descriptor.id)) {
                "Provider with ID \"${descriptor.id}\" is already registered."
            }
            providers[descriptor.id] = provider
        }
    }

    fun get(id: String): ProviderAdapter? = lock.withLock {
        providers[id]
    }

    fun list(): List<ProviderAdapter> = lock.withLock {
        providers.values.toList()
    }

    fun clear() {
        lock.withLock {
            providers.clear()
        }
    }
}
